server/core/cache.service.ts

```diff
@@ -0,0 +1,61 @@
+import { getAppSettings } from "./app.settings";
+
+interface Cache<T> {
+	[id: number]: T | null
+}
+
+export interface CacheStats {
+    cacheHits: number,
+    cacheMisses: number
+}
+
+const counts: CacheStats = {
+    cacheHits: 0,
+    cacheMisses: 0
+};
+
+export function createCache<T>() {
+    const cache: Cache<T> = {}
+
+    function get(id: number): T | null {
+        if (!getAppSettings().isCacheEnabled) {
+            return null
+        }
+
+        const cached = cache[id]
+        
+        if (cached) {
+            counts.cacheHits++
+        } else {
+            counts.cacheMisses++
+        }
+
+        return cached
+    }
+
+    function set(id: number, object: T) {
+        if (!getAppSettings().isCacheEnabled) {
+            return
+        }
+
+        if (cache[id] == null) {
```

Comment:
What if an object at the specified id changes its value? We wont be able to re-set it after the first time it's set (unless we clear it first) -- is this the intended behavior?

Reply:
We're supposed to invalidate the cached object using `clear` method in this class manually whenever that object changes. And then it's gonna get added to cache next time it's fetched.

That part is ok, but the question is how do we invalidate it?

> There are only two hard things in Computer Science: cache invalidation and naming things.

I'm still experimenting with where and how to use this (directly or automatically through some middleware), but in any way this `cache.service.ts` will probably remain the same.

Reply:
Right, thanks,